File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.js
const main = (arr) => {
    const dp = arr.map(() => 0)
    dp.push(0)
    
    for(let i = 1; i <= arr.length; i++){
        dp[i] = Math.max(dp[i], arr[i - 1]);
        for (let j = i + 1; j <= arr.length; j ++){
            dp[j] = Math.max(dp[j - i] + dp[i], dp[j])
        }
    }
    
    return dp[arr.length];
}

const fs = require('fs');
let [n, arr] = fs.readFileSync('/dev/stdin').toString().trim().split("\n");
arr = arr.split(" ").map(Number)

console.log(main(arr))

File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.py
def main(arr):
    dp = [0] * (len(arr) + 2)

    for i in range(1, len(arr) + 1):
        dp[i] = max(arr[i - 1], dp[i])
        for j in range(i + 1, len(arr) + 1):
            dp[j] = max(dp[j - i] + dp[i], dp[j])

    return dp[len(arr)]


n = int(input())
arr = [int(i) for i in input().split()]
print(main(arr))
